src/hotspot/share/services/nmt/memoryLogRecorder.cpp

```diff
@@ -0,0 +1,961 @@
+      size_t mark = MIN((size_t)round(-(1.0/(double)_histogram_horizontal_space)*overhead_ratio*overhead_ratio + 2.0*overhead_ratio), 100);
+
+      r_total += histogram_counts[i] * _malloc_requests_buckets[i];
+      a_total += histogram_counts[i] * histogram_actual_sizes[i];
+      o_total += overhead;
+
+      if (overhead_ratio > cutoff) {
+        if (overhead_ratio < 10.0) {
+          fprintf(stderr, "%10zu%c %9zu %9zu    %6zu  %02.3f ",
+                  _malloc_requests_buckets[i], flag, histogram_actual_sizes[i], histogram_counts[i], overhead, overhead_ratio);
+        } else {
+          fprintf(stderr, "%10zu%c %9zu %9zu    %6zu  %02.2f ",
+                  _malloc_requests_buckets[i], flag, histogram_actual_sizes[i], histogram_counts[i], overhead, overhead_ratio);
+        }
+        for (size_t j = 0; j < mark; j++) {
+          fprintf(stderr, "*");
+        }
+        for (size_t j = mark; j <= _histogram_horizontal_space; j++) {
+          fprintf(stderr, ".");
+        }
+        fprintf(stderr, "\n");
+      }
+    }
+  }
+
+  size_t actual_sizes_count = 0;
+  size_t actual_sizes_last = 0;
+  fprintf(stderr, "\n");
+  fprintf(stderr, "native malloc used following distinct allocation sizes: ");
+  for (size_t i = 0; i < _malloc_requests_count; i++) {
+    if ((histogram_actual_sizes[i] > 0) && (histogram_actual_sizes[i] > actual_sizes_last)) {
+      actual_sizes_count++;
+      actual_sizes_last = histogram_actual_sizes[i];
+      fprintf(stderr, "%zu ", histogram_actual_sizes[i]);
+    }
+  }
+  fprintf(stderr, "\n\n");
+  fprintf(stderr, "native malloc used %zu distinct allocation sizes\n", actual_sizes_count);
+
+//  assert(r_total == total_requested, "r_total:%zu == total_requested:%zu", r_total, total_requested);
+//  assert(a_total == total_actual, "a_total:%zu == total_actual:%zu", a_total, total_actual);
+//  assert(a_total-r_total == o_total, "a_total-r_total:%zu == o_total:%zu", a_total-r_total, o_total);
+//  assert(alloc_overhead == o_total, "alloc_overhead:%zu == total_requested:%zu", alloc_overhead, o_total);
+
+  ALLOW_C_FUNCTION(::free, ::free(histogram_actual_sizes);)
+  ALLOW_C_FUNCTION(::free, ::free(histogram_counts);)
+}
+
+void NMT_MemoryLogRecorder::print_entry(Entry* e) {
+  if (e != nullptr) {
+    fprintf(stderr, "{ %18ld, %18p, %18p", e->time, e->ptr, e->old);
+    for (int i = 0; i < NMT_TrackingStackDepth; i++) {
+      fprintf(stderr, ", %18p", e->stack[i]);
+    }
+    fprintf(stderr, ", %7u, %7u, %7u, %1d", (unsigned)e->requested, (unsigned)e->actual, (unsigned)e->flags, is_active(e));
+    fprintf(stderr, ", \"%s\"},\n", NMTUtil::flag_to_name(e->flags));
+  } else {
+    fprintf(stderr, "nullptr\n");
+  }
+}
+
+void NMT_MemoryLogRecorder::print_records(Entry* entries, size_t count) {
+  for (size_t c = 0; c < count; c++) {
+    Entry* e = &entries[c];
+    print_entry(e);
+  }
+}
+
+void NMT_MemoryLogRecorder::report_by_component(Entry* entries, size_t count) {
+  fprintf(stderr, "\n");
```

Comment:
Do you really have exclusive access to stderr when you print this? You could accumulate the string using `stringStream` and then print it all in one go instead.